07.ls_object/ls.rb

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+require './ls_file'
+require 'optparse'
+
+class LS
+  MAX_ROW = 3
+
+  def main
+    opt = OptionParser.new
+
+    params = { dot_match: false, long_format: false, reverse: false }
+    opt.on('-a') { |v| params[:dot_match] = v }
+    opt.on('-l') { |v| params[:long_format] = v }
+    opt.on('-r') { |v| params[:reverse] = v }
+    opt.parse!(ARGV)
+    path = '.'
+
+    files = collect_files(path, params[:dot_match], params[:reverse])
+    puts params[:long_format] ? list_long(files) : list_short(files)
+  end
+
+  private
+
+  def collect_files(path, dot_match, reverse)
+    filenames = Dir.foreach(path)
+    selected_filenames = dot_match ? filenames : filenames.reject { |file| file.start_with?('.') }
+    sorted_filenames = sort_filenames(selected_filenames)
+    reverse ? sorted_filenames.reverse : sorted_filenames
+  end
+
+  def sort_filenames(filenames)
+    filenames.sort do |a, b|
+      a = a.slice(1..-1) if a.start_with?('.') && !['.', '..'].include?(a)
+      b = b.slice(1..-1) if b.start_with?('.') && !['.', '..'].include?(b)
+      a.downcase <=> b.downcase
+    end
+  end
+
+  def list_long(filenames)
+    ls_files = filenames.map { |filename| LSFile.new(filename) }
+    block_total = ls_files.sum(&:blocks) / 2
+    total = "total #{block_total}"
+    body = format_body(ls_files)
+    [total, *body].join("\n")
+  end
+
+  def format_body(ls_files)
+    max_sizes = %i[nlink user group size].to_h do |key|
+      max = ls_files.map { |ls_file| key_size(ls_file, key.to_s) }.max
+      [key, max]
+    end
+    ls_files.map { |ls_file| format_row(ls_file, max_sizes) }
+  end
+
+  def key_size(ls_file, key)
+    case key
+    when 'nlink'
+      ls_file.nlink.to_s.size
+    when 'user'
+      ls_file.user.size
+    when 'group'
+      ls_file.group.size
+    when 'size'
+      ls_file.file_size.to_s.size
+    end
+  end
+
+  def format_row(ls_file, max_sizes)
+    [
+      "#{ls_file.type}#{ls_file.mode}",
+      ls_file.nlink.to_s.rjust(max_sizes[:nlink]),
+      ls_file.user.ljust(max_sizes[:user]),
+      ls_file.group.ljust(max_sizes[:group]),
+      ls_file.file_size.to_s.rjust(max_sizes[:size]),
+      ls_file.mtime.strftime('%b %e %H:%M'),
+      ls_file.name
+    ].join(' ')
+  end
+
+  def list_short(filenames)
```

Comment:
呼び出し側が `list_short(files)` なので、ここも
```suggestion
  def list_short(files)
```
と名前をあわせた方がベターです。